### PR TITLE
Avahi HA

### PR DESCRIPTION
**Pull request by vktg**
Redmine Issue: https://redmine.pfsense.org/issues/10165
Ready for review

HA support for Avahi package

### CODE REVIEW

File: net/pfSense-pkg-Avahi/files/usr/local/pkg/avahi.xml [hunk 1]
```diff
@@ -41,6 +41,11 @@
 		<executable>avahi-daemon</executable>
 		<description>Avahi mDNS/DNS-SD daemon</description>
 	</service>
+	<plugins>
```
Comment (jim-p) [inline net/pfSense-pkg-Avahi/files/usr/local/pkg/avahi.xml]: Indents must be tabs, not spaces.

File: net/pfSense-pkg-Avahi/files/usr/local/pkg/avahi/avahi.inc [hunk 2]
```diff
@@ -117,8 +117,29 @@ function avahi_sync_config() {
 		return;
 	}
 
-	log_error("Starting service avahi");
-	start_service('avahi');
+	/* Start/Restart Avahi, if a CARP VIP is set, check its status and act
+         * appropriately. */
```
Comment (jim-p) [inline net/pfSense-pkg-Avahi/files/usr/local/pkg/avahi/avahi.inc]: Indents must be tabs, not spaces (from here down there is a mix)

File: net/pfSense-pkg-Avahi/files/usr/local/share/pfSense-pkg-Avahi/info.xml [hunk 3]
```diff
@@ -7,5 +7,10 @@
 		<descr><![CDATA[Avahi is a system which facilitates host and service discovery in local networks via mDNS (Multicast DNS) and DNS-SD (DNS Service Discovery).]]></descr>
 		<version>%%PKGVERSION%%</version>
 		<configurationfile>avahi.xml</configurationfile>
+		<plugins>
```
Comment (jim-p) [inline net/pfSense-pkg-Avahi/files/usr/local/share/pfSense-pkg-Avahi/info.xml]: This is not necessary, the `<plugins>` section is not read from this file.

File: net/pfSense-pkg-Avahi/files/usr/local/www/avahi_settings.php [hunk 4]
```diff
@@ -158,6 +159,13 @@
 	$pconfig['enable']
 ));
 
+$section->addInput(new Form_Select(
+	'carpstatusvid',
+	'CARP Status IP',
```
Comment (jim-p) [inline net/pfSense-pkg-Avahi/files/usr/local/www/avahi_settings.php]: Should be `VIP` not `IP`

File: net/pfSense-pkg-Avahi/files/usr/local/www/avahi_settings.php [hunk 5]
```diff
@@ -158,6 +159,13 @@
...
+	$pconfig['carpstatusvid'],
+	avahi_get_carp_list()
+))->setHelp('Used to determine the CARP status. When the CARP vhid is in BACKUP status, Avahi will not be started.');
```
Comment (jim-p) [inline net/pfSense-pkg-Avahi/files/usr/local/www/avahi_settings.php]: Could use better wording:
`Used to determine the HA MASTER/BACKUP status. Avahi will be stopped when the chosen VIP is in BACKUP status, and started in MASTER status.'`

File: net/pfSense-pkg-Avahi/files/usr/local/pkg/avahi/avahi.inc (hunk 2 above)
Comment (vktg) [reply net/pfSense-pkg-Avahi/files/usr/local/pkg/avahi/avahi.inc]: fixed

File: net/pfSense-pkg-Avahi/files/usr/local/www/avahi_settings.php (hunk 5 above)
Comment (vktg) [reply net/pfSense-pkg-Avahi/files/usr/local/www/avahi_settings.php]: done

File: net/pfSense-pkg-Avahi/files/usr/local/www/avahi_settings.php (hunk 4 above)
Comment (vktg) [reply net/pfSense-pkg-Avahi/files/usr/local/www/avahi_settings.php]: done

File: net/pfSense-pkg-Avahi/files/usr/local/share/pfSense-pkg-Avahi/info.xml (hunk 3 above)
Comment (vktg) [reply net/pfSense-pkg-Avahi/files/usr/local/share/pfSense-pkg-Avahi/info.xml]: fixed

File: net/pfSense-pkg-Avahi/files/usr/local/pkg/avahi/avahi.inc [hunk 6]
```diff
@@ -130,3 +150,44 @@ function avahi_deinstall_command() {
...
 	unlink_if_exists(AVAHI_RCFILE);
 }
+
+function avahi_get_carp_list() {
+	$list = array();
+	$list['none'] = 'none';
+	$carplist = get_configured_vip_list("all", VIP_CARP);
+	foreach ($carplist as $vid => $vaddr) {
+		$vip = get_configured_vip($vid);
+		$list[$vid] = "$vaddr (vhid: {$vip['vhid']})";
```
Comment (jim-p) [inline net/pfSense-pkg-Avahi/files/usr/local/pkg/avahi/avahi.inc]: Inside double quotes, `$vaddr` should be surrounded by curly braces: `{$vaddr}`

File: net/pfSense-pkg-Avahi/files/usr/local/pkg/avahi/avahi.inc [hunk 7]
```diff
@@ -130,3 +150,44 @@ function avahi_deinstall_command() {
...
+	}
+	/* If there is no properly configured CARP status check IP, then stop */
+	if (!isset($avahi_config['carpstatusvid']) || $avahi_config['carpstatusvid'] == "none") {
+		return null;
+	}
+	/* Locate the VIP and match it against the configured check address */
+	list($vhid, $iface) = explode("@", trim($pluginparams['interface']));
+	$friendly = convert_real_interface_to_friendly_interface_name($iface);
+	$vip = get_configured_vip($avahi_config['carpstatusvid']);
+	if ($vip['vhid'] != $vhid || $vip['interface'] != $friendly) {
```
Comment (jim-p) [inline net/pfSense-pkg-Avahi/files/usr/local/pkg/avahi/avahi.inc]: The individual tests should be enclosed in parenthesis:
```
(($vip['vhid'] != $vhid) || ($vip['interface'] != $friendly))
```

File: net/pfSense-pkg-Avahi/files/usr/local/pkg/avahi/avahi.inc (hunk 6 above)
Comment (vktg) [reply net/pfSense-pkg-Avahi/files/usr/local/pkg/avahi/avahi.inc]: done

File: net/pfSense-pkg-Avahi/files/usr/local/pkg/avahi/avahi.inc (hunk 7 above)
Comment (vktg) [reply net/pfSense-pkg-Avahi/files/usr/local/pkg/avahi/avahi.inc]: done

**Comment by vktg** [comment]
fixed, squashed

most of the code comes from FRR package

**Comment by jim-p** [comment]
Would be better if the commit message was more meaningful, something like `Implement Avahi HA Tracking. Issue #10165` rather than a generic message that makes it difficult to pick out what it does in the commit log.

All commit messages should be [descriptive of what is being changed](https://docs.netgate.com/pfsense/en/latest/development/developer-style-guide.html#developer-rules) and [include issue number references](https://docs.netgate.com/pfsense/en/latest/development/referencing-tickets-in-commit-messages.html).

**Comment by vktg** [comment]
np
done

**Pull request merged**